### PR TITLE
Bpr

### PR DESCRIPTION
**Pull request by frankwhzhang**

### CODE REVIEW

File: paddle/fluid/operators/bpr_loss_op.cc [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        "Label_Pos",
+        "(Tensor), the tensor which represents the ground truth. It has the "
+        "same shape with 'X' except the last dimension. the last dimension "
+        "size is 1.");
+    AddOutput("Y",
+              "(Tensor, default Tensor<float>), a tensor whose shape is same "
+              "with 'X' except that the last dimension size is 1. It "
+              "represents the sequence bpr loss.");
+    AddComment(R"DOC(
+BprLoss Operator.
```
Comment (typhoonzero) [inline paddle/fluid/operators/bpr_loss_op.cc]: Better to give the full name and paper link.

File: paddle/fluid/operators/bpr_loss_op.h [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    return x;
+  }
+};
+
+template <typename DeviceContext, typename T>
+class BprLossOpKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    auto* x = ctx.Input<Tensor>("X");
+    auto* labels_Pos = ctx.Input<Tensor>("Label_Pos");
```
Comment (typhoonzero) [inline paddle/fluid/operators/bpr_loss_op.h]: labels_Pos -> `labels_pos`

and variable key:

("Label_Pos") -> ("LabelPos")

File: paddle/fluid/operators/bpr_loss_op.h [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    for (int i = 0; i < step_size; ++i) {
+      int lbl_pos = label_pos_data[i];
+      PADDLE_ENFORCE_GE(lbl_pos, 0);
+      PADDLE_ENFORCE_LT(lbl_pos, class_num);
+      int index_pos = i * class_num + lbl_pos;
+      T sum = static_cast<T>(0);
+      for (int j = 0; j < class_num; j++) {
+        if (j == lbl_pos) continue;
+        int index_neg = i * class_num + j;
+        sum += TolerableValue<T>()(-std::log(
```
Comment (typhoonzero) [inline paddle/fluid/operators/bpr_loss_op.h:72]: Maybe add some "TODO" notes for later performance optimizations, using blas or eigen. Need to find a way to adapt `TolerableValue` though.

File: paddle/fluid/API.spec [hunk 4]
```diff
@@ -32,13 +32,6 @@ paddle.fluid.BuildStrategy.ReduceStrategy.__init__ __init__(self: paddle.fluid.c
...
 paddle.fluid.create_lod_tensor ArgSpec(args=['data', 'recursive_seq_lens', 'place'], varargs=None, keywords=None, defaults=None)
 paddle.fluid.create_random_int_lodtensor ArgSpec(args=['recursive_seq_lens', 'base_shape', 'place', 'low', 'high'], varargs=None, keywords=None, defaults=None)
-paddle.fluid.DataFeedDesc.__init__ ArgSpec(args=['self', 'proto_file'], varargs=None, keywords=None, defaults=None)
-paddle.fluid.DataFeedDesc.desc ArgSpec(args=['self'], varargs=None, keywords=None, defaults=None)
-paddle.fluid.DataFeedDesc.set_batch_size ArgSpec(args=['self', 'batch_size'], varargs=None, keywords=None, defaults=None)
-paddle.fluid.DataFeedDesc.set_dense_slots ArgSpec(args=['self', 'dense_slots_name'], varargs=None, keywords=None, defaults=None)
-paddle.fluid.DataFeedDesc.set_use_slots ArgSpec(args=['self', 'use_slots_name'], varargs=None, keywords=None, defaults=None)
-paddle.fluid.AsyncExecutor.__init__ ArgSpec(args=['self', 'place'], varargs=None, keywords=None, defaults=(None,))
-paddle.fluid.AsyncExecutor.run ArgSpec(args=['self', 'program', 'data_feed', 'filelist', 'thread_num', 'fetch', 'debug'], varargs=None, keywords=None, defaults=(False,))
```
Comment (typhoonzero) [inline paddle/fluid/API.spec]: Maybe you have generate this file using an earlier version, please update this file.

File: paddle/fluid/operators/bpr_loss_op.cc [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+This operator belongs to pairwise ranking loss. Label_pos is the desired item.
+The loss at a given point in one seesion is defined as:
```
Comment (typhoonzero) [inline paddle/fluid/operators/bpr_loss_op.cc]: seesion -> session

File: paddle/fluid/operators/bpr_loss_op.h (hunk 2 above)
Comment (frankwhzhang) [reply paddle/fluid/operators/bpr_loss_op.h]: done

File: paddle/fluid/operators/bpr_loss_op.cc (hunk 1 above)
Comment (frankwhzhang) [reply paddle/fluid/operators/bpr_loss_op.cc]: done

File: paddle/fluid/operators/bpr_loss_op.h (hunk 3 above)
Comment (frankwhzhang) [reply paddle/fluid/operators/bpr_loss_op.h:72]: done

File: paddle/fluid/API.spec (hunk 4 above)
Comment (frankwhzhang) [reply paddle/fluid/API.spec]: done

File: paddle/fluid/operators/bpr_loss_op.cc [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+  void InferShape(framework::InferShapeContext* ctx) const override {
+    PADDLE_ENFORCE(ctx->HasInput("X"), "Input(X) should be not null.");
+    PADDLE_ENFORCE(ctx->HasInput("LabelPos"),
+                   "Input(LabelPos) should be not null.");
+    PADDLE_ENFORCE(ctx->HasOutput("Y"), "Output(Y) should be not null.");
+
+    auto x_dims = ctx->GetInputDim("X");
+    auto label_Pos_dims = ctx->GetInputDim("LabelPos");
```
Comment (gavin1332) [inline paddle/fluid/operators/bpr_loss_op.cc]: label_pos_dims

File: paddle/fluid/operators/bpr_loss_op.cc [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    int rank = x_dims.size();
+    PADDLE_ENFORCE_EQ(rank, label_Pos_dims.size(),
```
Comment (gavin1332) [inline paddle/fluid/operators/bpr_loss_op.cc]: label_pos_dims again, and correct all similar typos

**Review by gavin1332 — DISMISSED** [review]

**Review by typhoonzero — DISMISSED** [review]

File: python/paddle/fluid/layers/nn.py [hunk 8]
```diff
@@ -1348,6 +1349,33 @@ def cross_entropy(input, label, soft_label=False, ignore_index=kIgnoreIndex):
     return out
 
 
+def bpr_loss(input, label_pos):
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]:
```
def bpr_loss(input, label_pos, name=None):
```

maybe label, other than label_pos ? Why need postfix `pos`?

File: python/paddle/fluid/layers/nn.py [hunk 9]
```diff
@@ -1348,6 +1349,33 @@ def cross_entropy(input, label, soft_label=False, ignore_index=kIgnoreIndex):
...
+    """
+    Bayesian Personalized Ranking Loss Operator.
+
+    This operator belongs to pairwise ranking loss. LabelPos is the desired item.
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: LabelPos -> `label_pos`

File: python/paddle/fluid/layers/nn.py [hunk 10]
```diff
@@ -1348,6 +1349,33 @@ def cross_entropy(input, label, soft_label=False, ignore_index=kIgnoreIndex):
...
+    The loss at a given point in one session is defined as:
+    $Y[i] = -\frac{1}{N_{i}-1} * \sum_{0\le j<N_{i},~ j\neq Label[i]}\log(\sigma(X[i, Label[i]]-X[i, j]))$
+
+    Learn more details by reading paper <session-based recommendations with recurrent
+    neural networks>(https://arxiv.org/abs/1511.06939)
+
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py:1362]: Add comments for arguments, please refer others API, like fc.

File: python/paddle/fluid/layers/nn.py [hunk 11]
```diff
@@ -1348,6 +1349,33 @@ def cross_entropy(input, label, soft_label=False, ignore_index=kIgnoreIndex):
...
+    """
+
+    helper = LayerHelper('bpr_loss', **locals())
+    out = helper.create_variable_for_type_inference(dtype=input.dtype)
+    helper.append_op(
+        type='bpr_loss',
+        inputs={'X': [input],
+                'LabelPos': [label_pos]},
+        outputs={'Y': [out]})
+    return out
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py:1387]: Add unit testing in python/paddle/fluid/tests/unittests/test_layers.py

**Review by qingqing01 — COMMENTED** [review]
Only reviewed API.

**Review by qingqing01 — DISMISSED** [review]
LGTM for API.

**Review by qingqing01 — APPROVED** [review]

**Review by typhoonzero — APPROVED** [review]

**Review by gavin1332 — APPROVED** [review]

**Pull request merged**